Add SocketCAN functionality

garrettsummerfi3ld opened the pull request:
Implemented the Linux [SocketCAN](https://docs.kernel.org/networking/can.html) driver into CANBridge.

This is based off of the [`REVrobotics/CANBridge`](https://github.com/REVrobotics/CANBridge) [SocketCAN](https://github.com/REVrobotics/CANBridge/tree/SocketCAN) branch with lots of updates to compiling, driver code, and rewriting SocketCAN related code already existing to utilize the Linux kernel drivers.

willtoth committed:
Start of SocketCAN implementation

willtoth committed:
Simple method to list available drivers

willtoth committed:
Update README.md

garrettsummerfi3ld committed:
Merge branch 'main' into SocketCAN

Brings SocketCAN from upstream back to origin with already existing changes from main for workflows and other changes.

garrettsummerfi3ld committed:
Update SocketCANDevice class implementation

Some of the functions and implementations were not readily available nor created. This now resolves those issues.

garrettsummerfi3ld committed:
feat: Implement SocketCANDevice class

This commit adds the implementation for the SocketCANDevice class, which was previously missing some functions and implementations. Now all required functions are available and the issues have been resolved.

garrettsummerfi3ld committed:
Implement SocketCANDeviceThread in SocketCANDevice class

This was missing the previous commit.

Change compiler options to have SocketCANThread actually compile instead of not.

garrettsummerfi3ld committed:
refactor: Update include path for CanData header in SocketCANThread.h

Mirroring SerialDeviceThread.h as it was changed years ago.

garrettsummerfi3ld committed:
Fix compiler issues and broken functions

Adding more compiler args to prevent other platforms that are not able to compile some platforms to compile files, as it would just error and fail the build.

Fixing broken functions such as `SocketCANDevice::GetID()` as it was not returning a string. This is now not implemented but it returns an `int` which is correct for the function.

garrettsummerfi3ld committed:
Fix SocketCANDevice compilation and broken functions

Fix SocketCANDevice from being broken on compiles, from broken constructors to missing includes.

garrettsummerfi3ld committed:
Implement SocketCANDriver

This is largely untested, however this should bring the functionality of the SocketCAN work somewhat back into play.

garrettsummerfi3ld committed:
Rewrite SocketCANThread header

This tears out the majority of what was [candle_dll](https://github.com/HubertD/candle_dll/tree/master) for the Linux driver and implemented the Linux [SocketCAN](https://docs.kernel.org/networking/can.html) driver.

Preliminary testing shows that this does functionally work with a single SPARK MAX motor controller, however further testing may be required.

garrettsummerfi3ld committed:
Update README.md

Updated section for Linux support.

ItsHarper commented on src/main/native/cpp/CANBridge.cpp:
```diff
@@ -60,8 +64,12 @@ struct CANBridge_Scan {
 static const std::vector<rev::usb::CANDriver*> CANDriverList = {
 #ifdef _WIN32
     new rev::usb::CandleWinUSBDriver(),
+    new rev::usb::SerialDriver()
```
I'm not sure if this is Toth's old branch or a change you made, but please remove this line (as I've mentioned previously, I'd prefer to fully delete the old serial driver before we merge in cross-platform support)

garrettsummerfi3ld replied:
Removed mention from file, it still is in source however that can be removed at a later time.

ItsHarper commented on src/main/native/cpp/Drivers/SocketCAN/SocketCANDevice.cpp:
```diff
@@ -0,0 +1,118 @@
+SocketCANDevice::SocketCANDevice(std::string port) :
+    m_thread(port) {
+        m_descriptor = port;
+        m_name = "SPARK MAX";
```
We definitely don't want to hardcode this to "SPARK MAX". Ideally we'd get this information from the Linux kernel. I think the way to do this might be the `gs_usb_probe()` function (see what `usb_interface.dev.init_name` gives you).

If we can't figure that out and up needing to hardcode, use "USB CAN Device".

ItsHarper replied:
That function may not be available to you. I momentarily forgot that we're using the SocketCAN system, not the gs_usb kernel driver directly.

ItsHarper replied:
For now you can feel free to just hardcore it.

garrettsummerfi3ld replied:
I've changed it to be generic, but I don't think it really matters too much, marking this as resolved unless something else is missing.

ItsHarper reviewed (changes requested):
Reviewing this will probably take a while. Feel free to ping me if it's been a while since I've left any comments. It's good to see progress being made!

garrettsummerfi3ld committed:
Remove dangling serial driver

Plans to fully delete the serial driver are underway, as this is not supported or even functioning with current firmware.

garrettsummerfi3ld committed:
Change hardcoded name for SocketCANDevice

SocketCAN devices aren't exclusively, changing it to something more generic.
